slides/02-Using-gem5/04-cores.md

```diff
@@ -13,9 +13,643 @@ title: Modeling CPU cores in gem5
+  }
+</style>
+
+<div class="line">Simple CPU</div>
+
+---
+
+## SimpleCPU
+
+### **Atomic**
+
+Seq. of nested calls
+Use: Warming up, FF
+<!-- </br> -->
+
+### **Functional**
+
+Backdoor access to mem.
+(loading binaries)
+No effect on coherency states
+
+### **Timing**
+
+Split transactions
+Models queuing delay and
+resource contention
+
+![bg auto width:1250px Diagram to show different CPU Model Timings](04-cores-imgs/Simple-CPU.png)
+
+---
+
+## Other Simple CPUS
+
+
+### AtomicSimpleCPU
+
+- Uses **_Atomic_** memory accesses
+  - No resource contentions or queuing delay
+  - Mostly used for fast-forwarding and warming of caches
+
+### TimingSimpleCPU
+
+- Uses **_Timing_** memory accesses
+  - Execute non-memory operations in one cycle
+  - Models the timing of memory accesses in detail
+
+---
+
+## O3CPU (Out of Order CPU Model)
+
+- **_Timing_** memory accesses _execute-in-execute_ semantics
+- Time buffers between stages
+
+![width:1000px O3CPU](04-cores-imgs/O3CPU.svg)
+
+---
+
+## The O3CPU Model has many parameters
+
+[src/cpu/o3/BaseO3CPU.py](../../gem5/src/cpu/o3/BaseO3CPU.py)
+
+```python
+...
+decodeToFetchDelay = Param.Cycles(1, "Decode to fetch delay")
+renameToFetchDelay = Param.Cycles(1, "Rename to fetch delay")
+iewToFetchDelay = Param.Cycles(1, "Issue/Execute/Writeback to fetch delay")
+commitToFetchDelay = Param.Cycles(1, "Commit to fetch delay")
+fetchWidth = Param.Unsigned(8, "Fetch width")
+fetchBufferSize = Param.Unsigned(64, "Fetch buffer size in bytes")
+fetchQueueSize = Param.Unsigned(
+    32, "Fetch queue size in micro-ops per-thread"
+)
+renameToDecodeDelay = Param.Cycles(1, "Rename to decode delay")
+...
+```
+
+---
+
+## The O3CPU Model Parameters are easy to configure
+
+[src/cpu/o3/BaseO3CPU.py](../../gem5/src/cpu/o3/BaseO3CPU.py)
+
+In the following code, there are 256 physical integer registers
+
+```python
+    numPhysIntRegs = Param.Unsigned(
+        256, "Number of physical integer registers"
+    )
+```
+
+We can update the code so that there are only 32 physical integer registers
+
+```python
+    numPhysIntRegs = Param.Unsigned(
+        32, "Number of physical integer registers"
+    )
+```
+
+---
+
+## MinorCPU
+
+![bg auto width:700px Diagram to show different CPU Models](04-cores-imgs/Minor-CPU.png)
+
+<!-- 'https://nitish2112.github.io/post/gem5-minor-cpu/' Add "footer: " within the comment to make it appear on the slide-->
+
+---
+
+## KvmCPU
+
+- KVM – Kernel-based virtual machine
+
+- Used for native execution on x86 and ARM host platforms
+
+- Guest and the host need to have the same ISA
+
+- Very useful for functional tests and fast-forwarding
+
+---
+
+## Summary of gem5 CPU Models
+
+### **BaseKvmCPU**
+
+- Very fast
+- No timing
+- No caches, BP
+
+### **BaseSimpleCPU**
+
+- Fast
+- Some timing
+- Caches, limited BP
+
+### **DerivO3CPU and MinorCPU**
+
+- Slow
+- Timing
+- Caches, BP
+
+![bg width:1200px Diagram to show different CPU Models](04-cores-imgs/Summary-of-gem5-models-bg-2.png)
```

Comment:
These pictures are *very pretty*. Could you make them svgs, and editable with draw.io? The goal with these slides is to be easy to update in the future :smile: